### ISSUE TITLE
[Suggestion] connect_timeout option

### ISSUE THREAD

**Issue by Dillybob1992** [crystal-lang/crystal-mysql#51]
Hey! Awesome repo, and been using it for a couple days now and love it!
I have a small suggestion, that I think could be beneficial. A `connect_timeout` property, that can be passed in the query string

For example:
```
  def connect_to_mysql(login, password, database)
    DB.open "mysql://#{login}:#{password}@localhost/#{database}?connect_timeout=3"
  rescue e
    puts "MYSQL Connection Error: #{e.message}"
    exit
  end
```

This way, the connection won't "hang" on localhost, if no MYSQL server is running

**Comment by bcardiff** [comment]
That should be already covered by the current pool options. Check https://crystal-lang.org/docs/database/connection_pool.html .

If there is no database available it should not hang. Probably something else is hanging the execution of you app. Where you able to perform the `DB.open` in an isolated manner and experience the "hang"?

**Comment by Dillybob1992** [comment]
@bcardiff Hi!

Here is some isolated code
```
require "db"
require "mysql"

def connect_to_mysql(login, password, database)
  DB.open "mysql://#{login}:#{password}@localhost/#{database}?retry_attempts=0&retry_delay=0"
  puts "If the MYSQL server is online, and the password/user/databse is wrong, it will raise"
rescue e
  puts "MYSQL Connection Error: #{e.message}"
  exit
end

connect_to_mysql("root", "fakepassword", "master")

puts "This line never get printed, and app hangs here if MYSQL is offline"

```

Tried with `?retry_attempts=0&retry_delay=0` as well

**Comment by Dillybob1992** [comment]
Oops, my bad. Yeah, remove the `exit`. But the `ConnectionRefused` error somehow isn't being rescued, it just hangs there. That's not normal right?

**Comment by bcardiff** [comment]
There is some misconception here I think.

`DB.open` generates an exception. Since there is a `begin/rescue` the exception is caught and the rescue block is executed. The semantic is that the exception will be swallowed, unless there is a `raise e` inside the rescue to re-raise the exception.

I thought that by hangs you meant that the program freezes, but now I am questioning that assumption :-P

Ref: https://crystal-lang.org/docs/syntax_and_semantics/exception_handling.html

I'll close the issues since the original cause is solved. Don't miss the IRC for better answers regarding language semantics.

**Comment by Dillybob1992** [comment]
> the exception is caught and the rescue block is executed.

It's not, when the MYSQL server is offline. Is what i'm trying to say :P

**Comment by bcardiff** [comment]
Oh, ok. Then is probably something in Windows/bash (or maybe even linux 🤔 )
Could you check what kind of exception are you getting at instead of an `Errno` https://github.com/crystal-lang/crystal-mysql/blob/master/src/mysql/connection.cr#L31 ?

**Comment by Dillybob1992** [comment]
Haha, am not sure. It doesn't give any exception / rescue doesn't get triggered at all (in fact, anything right below the DB.open line doesn't print).  

However, when the app is hanging, and when I turn on my MYSQL server the `e.class` prints:
`Exception`

**Comment by Dillybob1992** [comment]
Did some testing in that `src/mysql/connection.cr`
```
      puts "i get printed"
      @socket = TCPSocket.new(host, port)
      handshake = read_packet(Protocol::HandshakeV10)

      puts "will never be printed"
```
Looks like something is happening here `handshake = read_packet(Protocol::HandshakeV10)`, where it just hangs. Could isolate it further but not sure where else to look

**Comment by Dillybob1992** [comment]
@bcardiff Yay! `connect_timeout: 5` didn't work but `connect_timeout: 1` did. 
e.class is
```
IO::Timeout
MYSQL Connection Error: connect timed out
```

**Comment by bcardiff** [comment]
That's odd. I am not confident the 1s is long enough, but at least is narrowed now. Thanks!

**Comment by Dillybob1992** [comment]
Yeah, that's weird. Anything above 1 doesn't raise the exception. Just tried 2,3,4 and 5

**Comment by bcardiff** [comment]
In linux at least is seems to not be an issue: I added a spec in a branch https://github.com/crystal-lang/crystal-mysql/commit/81d6ebba416261a085b53a7eb7c7a8f6f891ba3d and travis passed https://travis-ci.org/crystal-lang/crystal-mysql/builds/329689570
